0010-endpoint-metadata-extension/README.md

```diff
@@ -0,0 +1,687 @@
+---
+name: Endpoint Metadata Extension
+shortname: 10/EPMDX
+status: draft
+editor: Volodymyr Tkhir <vtkhir@cybervisiontech.com>
+---
+
+- [Introduction](#introduction)
+- [Requirements and constraints](#requirements-and-constraints)
+  - [Metadata keys](#metadata-keys)
+  - [Metadata values](#metadata-values)
+  - [Supported operations](#supported-operations)
+- [Design](#design)
+  - [Full metadata update](#full-metadata-update)
+    - [Full metadata update request](#full-metadata-update-request)
+    - [Full metadata update response](#full-metadata-update-response)
+  - [Partial metadata update](#partial-metadata-update)
+    - [Partial metadata update request](#partial-metadata-update-request)
+    - [Partial metadata update response](#partial-metadata-update-response)
+  - [Get metadata](#get-metadata)
+    - [Get metadata request](#get-metadata-request)
+    - [Get metadata response](#get-metadata-response)
+  - [Get metadata keys](#get-metadata-keys)
+    - [Get metadata keys request](#get-metadata-keys-request)
+    - [Get metadata keys response](#get-metadata-keys-response)
+  - [Delete metadata key](#delete-metadata-key)
+    - [Delete metadata key request](#delete-metadata-key-request)
+    - [Delete metadata key response](#delete-metadata-key-response)
+
+
+## Introduction
+
+The Endpoint Metadata Extension (EPMDX) protocol is an endpoint-aware [Kaa Protocol](/0001-kaa-protocol/README.md) extension.
+
+It is intended to manage endpoint metadata.
+
+## Requirements and constraints
+
+### Metadata keys
+EPMDX protocol supports only string type of metadata keys.
+All UTF-8 characters are supported in metadata key names.
+Metadata key must contain at least one non-whitespace character.
```

Comment:
> with empty metadataKey is api/v1/endpoints/{endpointId}/metadata/

In the EPR REST API we have two methods:

- `PUT api/v1/endpoints/{endpointId}/metadata` - updates all metadata
- `PUT api/v1/endpoints/{endpointId}/metadata/{metadataKey}` - updates single metadata key

So update of single metadata key URI should look exactly ` api/v1/endpoints/{endpointId}/metadata//`

> let's take ones from NATS (but probably forbid dots)

I agree on your proposal.
@akokhanovskyi what do you think about restriction of metadata key with following regex `^[a-zA-Z0-9]+$`?